### PR TITLE
feature: method to get reference to tuple elements satisfying type_trait

### PR DESCRIPTION
**Pull request by johnbowen42**
For context, this is a function I wrote for https://github.com/LLNL/RAJA/pull/1799, where we need to easily extract reference to kernel parameters that are reducers.  The interface requires passing a `tuple` typed object that the user desires to extract elements from, as well as a class-template that inherits from `std::true_type` for desired elements. Here are some alternatives I considered:
- Adding ability to return pointer to element, copy of element, or reference to element.  Right now the method only returns references to elements
- Instead of the type-trait interface, passing the template class directly as a parameter to the method.  This has a number of pitfalls, including the case where the desired template class itself is dependent upon other instantiations of template classes for its definition

I am not committed to the name `get_refs_to_elements_by_type_trait` and would like to hear alternative ideas

### CODE REVIEW

**Comment by trws** [comment]
Interesting, not a bad idea. Would you be willing to add a link to where it's used in that PR? I'm curious about the usage pattern, the implementation looks easy to follow, which is great, but we might be able to make it a bit faster on template expansion depending on the usage.

**Comment by johnbowen42** [comment]
> Interesting, not a bad idea. Would you be willing to add a link to where it's used in that PR? I'm curious about the usage pattern, the implementation looks easy to follow, which is great, but we might be able to make it a bit faster on template expansion depending on the usage.

@trws here https://github.com/LLNL/RAJA/blob/c8f02a9f7c1465d7d019176bb631ea5c65b1f9f8/include/RAJA/pattern/params/forall.hpp#L33

I changed my PR to reflect these changes, but this will break CI for the RAJA PR.  Also, we will need to rewrite camps' git history a little bit and bring in this PR prior to the cxx17 change.

File: include/camp/tuple.hpp [hunk 1]
```diff
@@ -624,6 +624,62 @@ CAMP_HOST_DEVICE constexpr auto tuple_cat_pair(L&& l, R&& r) noexcept
...
+{
+  return camp::tuple<> {};
+}
+
+template<camp::idx_t param_size,
+         template <typename> typename TypeTrait,
+         typename TupleType,
+         camp::idx_t idx>
+CAMP_HOST_DEVICE constexpr auto
+get_refs_to_elements_by_type_trait_impl(TupleType& param,
```
Comment (MrBurmark) [inline include/camp/tuple.hpp:666]: Do you want to put these impl functions in a detail namespace?

**Review by MrBurmark — DISMISSED** [review]

**Review by artv3 — DISMISSED** [review]
Looks good to me thanks @johnbowen42 !

File: test/tuple.cpp [hunk 2]
```diff
@@ -416,6 +416,28 @@ TEST(CampTaggedTuple, MakeTagged)
...
+
+template<typename T>
+struct IsSearchType : std::false_type {};
+
+template<typename Param>
+struct IsSearchType<SearchType<Param>> : std::true_type {};
+
+TEST(CampFilterByTypeTrait, MakeFilteredTuple)
+{
+  using BaseTupleType = camp::tuple<double, SearchType<int>, int, OtherType<int>, camp::tuple<double>, SearchType<int>>;
```
Comment (MrBurmark) [inline test/tuple.cpp:433]: Do you want to add some more tests like empty tuple, one element tuple, search type in the left most, in the right most, etc?

**Review by rhornung67 — APPROVED** [review]

**Review by MrBurmark — APPROVED** [review]

**Review by artv3 — APPROVED** [review]

**Comment by MrBurmark** [comment]
I assume @trws is talking about either reducing the number of function templates involved and/or using a binary tree kind of approach to reduce the number of recursive function calls.

**Comment by trws** [comment]
This is fine as is, I was hoping we could use an iterative rather than recursive approach for it, but what's coming in doesn't allow for an easy way to do that without a bunch of other infrastructure so it probably isn't worth it right now.

**Comment by johnbowen42** [comment]
> This is fine as is, I was hoping we could use an iterative rather than recursive approach for it, but what's coming in doesn't allow for an easy way to do that without a bunch of other infrastructure so it probably isn't worth it right now.

When we move to cxx17 this should be relatively simple with a fold expression, without building up much infrastructure

**Comment by trws** [comment]
I was thinking so too, but it isn't entirely straightforward because there's no good way to have a fold skip elements or filter into an argument list, especially using emulated folds in 14.  Definitely something to think about when we can though.
